packages/feathers/src/router.ts

```diff
@@ -63,71 +116,141 @@ export class RouteNode<T = any> {
+              this.placeholders.splice(i, 1)
+            }
+            break
+          }
+        }
+      }
     } else if (this.children[current]) {
       const child = this.children[current]
 
       child.remove(path)
 
-      if (!child.hasChildren) {
+      if (!child.hasChildren && child.data === undefined) {
         delete this.children[current]
       }
     }
   }
 
   lookup(path: string[], info: LookupData): LookupResult<T> | null {
-    if (path.length === this.depth) {
-      return this.data === undefined
-        ? null
-        : {
-            ...info,
-            data: this.data
-          }
+    // Early exit optimization - check for exact match at current depth
+    if (this.depth === path.length) {
+      if (this.data !== undefined) {
+        info.data = this.data
+        return info as LookupResult<T>
+      }
+
+      // Check for catch-all at this level (handles empty catch-all)
+      if (this.catchAll) {
+        info.data = this.catchAll.data
+        info.params[this.catchAll.name] = []
+        return info as LookupResult<T>
+      }
+
+      return null
     }
 
     const current = path[this.depth]
-    const child = this.children[current]
 
+    // Try exact child match first (most common case)
+    const child = this.children[current]
     if (child) {
-      const lookup = child.lookup(path, info)
-
-      if (lookup !== null) {
-        return lookup
+      const result = child.lookup(path, info)
+      if (result !== null) {
+        return result
       }
     }
 
-    // This will return the first placeholder that matches early
-    for (const placeholder of this.placeholders) {
+    // Only check placeholders if exact match failed
+    const placeholdersLength = this.placeholders.length
+    for (let i = 0; i < placeholdersLength; i++) {
+      const placeholder = this.placeholders[i]
       const result = placeholder.lookup(path, info)
-
       if (result !== null) {
         result.params[placeholder.name] = current
         return result
       }
     }
 
+    // Check catch-all as final fallback
+    if (this.catchAll) {
+      const remaining = path.slice(this.depth)
+      info.data = this.catchAll.data
+      info.params[this.catchAll.name] = remaining
+      return info as LookupResult<T>
+    }
+
     return null
   }
 }
 
-export class Router<T = any> {
+export class Router<T = any> implements RouterInterface<T> {
   public caseSensitive = true
+  private pathCache: { [key: string]: string[] } = Object.create(null) // Cache for parsed paths
 
   constructor(public root: RouteNode<T> = new RouteNode<T>('', 0)) {}
 
   getPath(path: string) {
-    const result = stripSlashes(path).split('/')
+    // Check cache first
+    if (this.pathCache[path]) {
+      return this.pathCache[path]
+    }
+
+    // Fast path for root
+    if (!path || path === '/') {
+      const result = ['']
+      this.pathCache[path] = result
```

Comment:
`pathCache` is never cleaned up so anybody can crash the server by making a lot of unique URL requests.

Reply:
Good catch.  I added a simple LRU mechanism that defaults to 1000 cached paths before booting old ones.